### PR TITLE
[CR] New mutation category: Mouse 🐁

### PR DESCRIPTION
**Pull request by ghost**
I had this idea in the morning and whipped it up over several sessions throughout the day.
* Want to stuff tons of junk food down your face and not get unhealthy?
* Want to sprint at the speed of sound at half the size of a normal human?
* Want to be a tiny terror with a katana that stabs ankles?
* Want to never drink anything, ever?
* Want every animal to hate you and want to end your existence?
* Want to dodge every hit if you don't have armor on?
* Want to die in three hits if you *don't* dodge them?

If you said 'yes' to any of these questions, Mouse might be for you.

# Short/spoiler-free version:

This adds a new mutation category called Mouse. Find the recipe for the mutagen in the Dionne journal, and make it with cheese as the special ingredient. It focuses on heavily increased stealth, speed, dodge, resourcefulness, and getting more out of what you have, with a tradeoff of less carrying capacity, higher encumbrance, increased hunger, and a large hit point reduction until you've gone post-threshold. Mechanically, it's a bit like a crossover between Rat and Bird - borrowing many of the same mutations from the former, and the increased speed from the latter.

# Long version:

Buckle up, this is a long list of things.

### Mutagen info:

* Craftable with cooking 9 and first aid 1. Recipe is learned from lab journal-Dionne, which has all the other animal mutagen recipes.
* Special ingredient is 2 of any cheese.
* Can be found naturally in serum form within labs.

#### Traits

* Fleet-Footed
* Quick
* Fast Metabolism
* Deft
* [Gourmand](https://i.imgur.com/p8cy0fO.jpg)
* Sleepy
* Scout
* Accomplished Sleeper
* Indefatigable
* Forgetful
* Addictive Personality
* Strong Scent
* Jittery
* Albino
* Flimsy

#### Mutations

* Lightly Furred
* Sleek Fur
* Light Bones
* High Night Vision
* Rat Tail (now renamed to Rodent Tail)
* Wiry Whiskers
* Rapid Metabolism (no longer purifiable)

#### Post-Mouse Threshold Mutations

* Eater Of The Dead
* Very Dextrous

#### Pre-Threshold

* Mouse Ears - Doubles the volume of heard sounds
* Little - Reduces hit points by 5% and weight capacity by 20%, but increases base speed by 5% and movement speed by 10%
* **Tiny** - This is *by far* biggest change for Mouse characters. It:
  * Halves weight capacity
  * Reduces all hit points by 30%
  * Silences movement
  * Increases dodge skill by 2
* Sweet Tooth - Grants a morale bonus from eating junk food (+5 for half an hour, stacking to +30)
* Prey Animal - Animal Discord, but they hate you and want you to die

#### Post-Mouse Threshold (named Diminutive)

* **Unassuming** - The post-threshold version of Tiny.
  * Penalties are reduced to represent being used to the new form
  * Weight capacity reduced by 25% instead of half
  * Hit points reduced by 25% instead of 30%
  * Retains all other changes of Tiny
  * Not purifiable
* Sugar-Loving - Upgrade of Sweet Tooth (+10 morale for an hour, stacking to +50)
  * Not purifiable
  * Also prevents food from being unhealthy to you
* Hyperactive - Upgrade of Indefatigable, increases max stamina by 40%
* Crafty - Reduces the range at which enemies detect you by 40%, and increases all speed by 15%, and movement speed by 25%
  * Not purifiable
* Practiced Sleeper - Upgrade of Accomplished Sleeper, lets you sleep nearly anywhere
* **Metabolic Rehydration** - Mice don't need to drink, and neither do you! Removes the need to drink water

### New mechanic: Undersizing

Being a mouse is tuff stuff, and your clothes are too big to fit you comfortably, resulting in doubled encumbrance if you have the Tiny or Unassuming mutations. However, there's a solution: undersize them! You can refit clothes that already fit you to undersize them if you're very small, which makes them fit you normally again and gives them an `(undersize)` tag in their name.

If you're normally-sized, you can reverse the undersizing so they fit normally again, and you really should, because wearing undersized clothes *triples* their encumbrance if you're not abnormally small like a mouse mutant.

## Rationale/Design Notes:

~~I made this entirely because I want to be a cute mouse person.~~

~~I made this entirely so people could cosplay Nazrin.~~

This morning, while still waking up, I had a realization - there are a decent chunk of mutations based around being abnormally large... but none based around being abnormally small. The core thematics of Mouse are based around that fact: being abnormally small. From this, other things stemmed - speed, deftness, and stealth, as well as the realistic drawbacks of being two feet tall - warm clothes are now REALLY warm, and things that used to fit you are now super baggy.

[Not only are mice small](https://www.urbandictionary.com/define.php?term=You%20don%27t%20say), they also [eat a lot](https://animals.mom.me/wild-mice-need-eat-11634.html), and [sleep a lot](https://medicalxpress.com/news/2007-03-mice-longer-humans.html). As such, Rapid Metabolism was moved into their mutations to represent this fact. Hyperactive was added to represent how their bodies are like tiny little suns. They also [prefer sweet food](https://animals.howstuffworks.com/animal-facts/do-mice-really-love-cheese.htm), so making them enjoy junk food more made sense and provides an interesting reason to hold onto the stuff later on. Finally, [they can survive without liquid water at all](http://icwdm.org/handbook/rodents/HouseMice.asp), so you don't need to drink.

Going post-threshold Mouse allows you to significantly reduce the drawbacks of being `supersmol.` This is to both give a reason to go post-threshold and deal with what drawbacks remain, and to show an adaptation and comfort with the new tiny body. A mouse mutant who feels at home in their small form would be able to use it as such, instead of having to relearn how to function.

Mouse's biggest upside is its speed increase; a full post-threshold Mouse character could achieve upwards of +30% speed and even more movement speed, and have a massive 40% extra stamina to make use of it.

Their two most massive drawbacks are the increased encumbrance from armor (+50% is no laughing matter at post-threshold, and +100% is staggering) and decreased carry weight (the combination of Light Bones and Tiny could reduce the base 90-lb weight into as little as *30 pounds*.)

Combined with their increased dodge ability, Mouse rewards players who travel as light and breezy as possible: hit-and-run attacks, as well as outrunning opponents, becomes far easier with post-threshold Mouse, but a Mouse who packs turnout gear and a crowbar when they start to fight a hulk may just find themselves with no hole to run back to with their doubled encumbrance and reduced health.

Mutation ideas, comments, feedback, and code suggestions are all highly welcomed and encouraged. This is a sizable addition, and the largest *gameplay* I've done for CDDA (though on the technical scale, I've made larger), so I'd love to shape it around what people would enjoy.

### CODE REVIEW

**Comment by DracoGriffin** [comment]
Compiled and tested. Works. At initial review, it seems incredibly favored towards positive mutations but once tested in-game, the Small mutation tree is much more punishing than I realized.

The mouse survivor might be quite agile, but the lack of weight to carry supplies or wear proper gear is a valid concern. A lot of the traits feel appropriate and suit the mutation tree, but my main concern is there doesn't seem to be a lot of negative traits, or particularly severe given the benefits, especially when paired with another tree that gives any Strength boost mutations (like Lupine).

Overall, I look forward to playing both a weak ranged plucky mouse with a pneumatic crossbow and a swashbuckling sword fighter mouse. I think it really requires 3-4 more negative traits (not sure what specifically yet, potentially hoarder as one) as otherwise, it's a considerably strong combat build. Also, it'd be neat if at Sugar-Loving, junk food stopped giving negative health.

Lastly, some of the new traits might be good to look at adding to other categories, potentially.

**Comment by Vasyan2006** [comment]
Most mutations change only one or two characteristics, but ´tiny´ change a lot. It will be better to separate it and use existed traits from bird, like bad back or hollow bones. 
No strength penalty? As I see in descriprion this character can not have more then 4. I played characters without ability to wear heavy armor or use heavy weapon - it is much more fun compare to common "hulk smash" gameplay with wearing 200 lbs of loot.
HP must be reduced more then by 10% to compensate overpowered dodge, otherwise it is not a mouse, but minibrute-troglobite.
Add more mouse-kin to "Mutant NPCs" mod.

**Comment by cainiaowu** [comment]
What is the difference between rat and mouse? I assume they are both rodent type of creature.

**Comment by MT-Arnoldussen** [comment]
I agree that Tiny and Small are rather insular mutations that would make more sense if they were a combination of several existing ones, like Bad Back, Flimsy, and probably a new Strength reducing one as well.

**Comment by Xhuis** [comment]
@cainiaowu Rat is more focused around being a "survivor" type, scrounging up resources from every possible spot, and staying under the cover of dark. Mouse is more focused around being a "fast and flimsy" type, moving very quickly but unable to stack armor due to increased encumbrance and warmth, getting the most out of traveling light and carrying little.

**Comment by Vasyan2006** [comment]
If postthreshold character is twice smaller consider new item flag "tiny". Almost every cloth and armor can be refitted and reduced in size (and will weight only 2/3 of initial), so will have normal encumbrance. But characters without "tiny" mutation will be unable to wear it.
As opposite every cloth item can be refitted and get "oversize" flag for huge characters. But it will require additional components and will increase item weight twice.

**Comment by DracoGriffin** [comment]
>What is the difference between rat and mouse? I assume they are both rodent type of creature.

Probably similar to Lupine/Feline/Beast, but with more significant divergence to actually be separate, but similar trees. Lupine/Feline/Beast should get a similar treatment in the future (most mutation trees probably should as well).

**Comment by kevingranade** [comment]
Awesome set of changes, you have a clear vision for what you want to happen, *and* a great angle on making it balanced.

I agree with @Vasyan2006 that the tiny and unassuming mutations are handling too many things at once, some of them don't seem to be related to being tiny either, such as increased speed (though they ARE related to "being a mouse") and being able to ignore mutation limits on clothing.

Speaking of which, disabling mutation/clothing interactions is not ok, if your mouse character also happens to have antlers or wings or a big tail, that's still going to make them unable to wear conflicting clothing, no matter how small they are.  If your mutation tree is going to be capable of wearing most clothing, it simply needs to avoid mutations that conflict with clothes, which as far as I can tell, this one does.

I don't follow the logic of the warmth increase from being small, *some* clothes are going to "bunch up" and provide effectively more insulation, but most are simply not going to fit properly and not provide meaningfully different insulation.

I also think that the penalties from oversized clothing can be negated by tailoring.

Implementation-wise, you may be able to get a lot of the effects you are overriding manually by changing the value returned by player::get_size()  (though ideally this would also be hoisted to the character class), smaller effective size is already handled in many places due to variable-sized monsters, and we can leverage many of those changes for mutant characters as well.

**Comment by Xhuis** [comment]
My desktop is not functioning, so I'm leaving this comment as a reminder for myself for what needs to be done.

* Split Tiny and Unassuming into multiple mutations
* Undersizing clothing, changing around warmth values
* More negative traits
* Mutation clothes tweaking
* Modifying player mob size
* Reduce stat boost mutations and dodge bonuses

**Comment by MT-Arnoldussen** [comment]
I would suggest Jittery for a negative trait. Seems fitting for a mouse.

**Comment by Cyrano7** [comment]
I think mouse ears and rat ears are pretty similar and could be called rodent ears and given to both.

**Comment by Cyrano7** [comment]
How about a mutation for needing to drink less water? Mice don't need much and can get enough water from just the food they eat sometimes. Especially certain species.

**Comment by DracoGriffin** [comment]
>How about a mutation for needing to drink less water? Mice don't need much and can get enough water from just the food they eat sometimes. Especially certain species.

I don't know about all mice (are we looking at species or genus level or even higher?) but I believe I posted this somewhere else about [kangaroo rats](https://en.wikipedia.org/wiki/Kangaroo_rat#Range_and_habitat):
>They do this in part by lowering their metabolic rate, which reduces loss of water through their skin and respiratory system. They obtain enough water from the metabolic oxidation of the seeds they eat to survive and do not need to drink water at all.

**Comment by Xhuis** [comment]
That is a very good point. I'll put that on the todo.

**Comment by Xhuis** [comment]
Since my desktop appears to be out of commission, I bit the bullet and did some work on this with my laptop.
* Adds a new `UNDERSIZE` flag for clothing. Small mutants will have double encumbrance for any non-undersized clothing, but will have normal encumbrance if it is undersized. Normal people will have *triple* encumbrance for undersized clothing.
  * `UNDERSIZE` is added and removed by refitting clothing. It will be added or removed automatically depending on if you're small or not. You can refit clothes that already fit you if they're not your size.
* Sugar-Loving now sets max unhealthiness of food to 0, since your body is built around eating junk.
* Warmth isn't doubled by clothing anymore, since you can undersize it.
* Tiny and Unassuming dodge bonuses reduced and placed into Crafty. Speed bonuses removed and also placed into Crafty.
* Unassuming drops max HP by 25%, instead of 10%.
* Rats can have Rodent Ears too.
* Adjusted the mutations list.
  * **New traits for mice:** Forgetful, Addictive Personality, Strong Scent, Jittery, Albino
  * **Removed traits for mice:** Substance Tolerance, Light Step, Weak Scent, 
  * **New pre-thresh muts:** Prey Animal (stronger version of Animal Discord)
  * **New post-thresh muts:** Eater Of The Dead, Very Dextrous (instead of Insanely Dextrous), Metabolic Rehydration (removes need to drink water)
  * **Removed post-thresh muts:** Extremely Perceptive

I aimed to give a *lot* more negatives, while still keeping them powerful and worthwhile post-threshold with new, interesting mutations. Updating OP as we speak.

**Comment by Xhuis** [comment]
Yes.

**Comment by Vasyan2006** [comment]
> Reduce stat boost mutations and dodge bonuses

In this case it will be "just another beast-like mutant". Every mutant category must be unique.
I suggest to keep highest dodge value, but add more 
> Prey Animal

mutations. Something like "pacifist2" trait: all combat skill (except dodge) will grow at 10% rate. This nonviolent resistance will make direct combat (or will try to make it) mostly pointless. 

Further evolution of "Sugar-Loving" - any non-junk food add huge morale debuff. Mouse character will be forced to follow strict diet rules.

Just some ideas I want to see ingame.

**Comment by AMurkin** [comment]
May mouse pass through broken window like through clear tile? For the benefit of this mutation.

**Comment by kevingranade** [comment]
There is an effect that lets sufficiently small (I think tiny?) monsters
avoid damage from hazardous terrain.

**Comment by FantasticFwoosh** [comment]
level 10 mutant cooking ability

Looks nice, but from the title i was expecting a more Skaven like man-rat creature thriving in squalor. Not dissapointed though on the new mechanics.

**Comment by Rivet-the-Zombie** [comment]
>Looks nice, but from the title i was expecting a more Skaven like man-rat creature thriving in squalor able to chew its way through walls and furniture.

We already have that in the form of the rat mutation tree.

**Comment by Xhuis** [comment]
As far as the scope of this PR, I believe I've done everything I wanted to do. I know reducing the player's mob size has come into question a good bit, so I will seek investigating that, but I don't know the code hierarchy enough to whip it up real quick, so it may take some time to replace every instance of the code.

I may also simply make the current size check a wrapper for the new size check, so as to allow legacy support in existing code and not have to sink the PR's lines into a lot more files.

**Comment by DracoGriffin** [comment]
Double-check astyle on .cpp changes

**Comment by Xhuis** [comment]
Any possibility of this getting merged soon?

**Comment by Zireael07** [comment]
merging this PR seems to have caused all further pr builds to stall?

File: src/consumption.cpp [hunk 1]
```diff
@@ -801,8 +815,13 @@ void player::consume_effects( const item &food )
...
     mod_stomach_water( comest.quench );
-    if( comest.healthy != 0 ) {
+    int effective_health = comest.healthy;
+    if( effective_health != 0 ) {
         // Effectively no cap on health modifiers from food
+        if( has_trait( trait_id( "PROJUNK2" ) ) && effective_health < 0 ) {
+            effective_health =
+                0; // Our digestive system is built around crappy food, so it can handle it just fine
+        }
         mod_healthy_mod( comest.healthy, ( comest.healthy >= 0 ) ? 200 : -200 );
```
Comment (Qrox) [inline src/consumption.cpp]: Here `effective_health` is changed but this line still uses comest.healthy, so it will not have the intended effect.

**Comment by Xhuis** [comment]
I arrived home early from my camping trip, so I can try and look at this. Do we have any idea what might be causing the stalling?

**Comment by kevingranade** [comment]
Looks like something is causing the test suite to go into an infinite loop.

**Comment by Xhuis** [comment]
That's bizarre. Why might that be?

**Comment by kevingranade** [comment]
I haven't had a chance to dive in yet, but I suspect some assumption of the test is being invalidated.

It's almost certainly happening here https://github.com/CleverRaven/Cataclysm-DDA/blob/9f7c1acdd5d95787aa5047794f8334ab7dc4ac0b/tests/mutation_test.cpp#L21

It looks like it's trying to reach every mutation in the mouse category, but its failing to ever reach one of them.

Should be able to just instrument that test to print what it's trying to do and figure out what's happening.

**Comment by Xhuis** [comment]
Simple enough. I'm not familiar with how testing works so I'll do some research into it.

**Pull request merged**